### PR TITLE
LISI implementation

### LINKED ISSUE #9: Implement lisi

### PR DESCRIPTION
**Pull request by justjhong**
Fixes #9

### CODE REVIEW

File: src/scib_metrics/utils/_lisi.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    def _get_neighbor_probability_step(state):
+        _, _, Hdiff, beta, betamin, betamax, tries = state
+        new_betamin = jnp.where(Hdiff > 0, beta, betamin)
+        new_betamax = jnp.where(Hdiff > 0, betamax, beta)
+        new_beta = jnp.where(Hdiff > 0, jnp.where(betamax == jnp.inf, beta * 2, (beta + betamax) / 2),
+                             jnp.where(betamin == -jnp.inf, beta / 2, (beta + betamin) / 2))
+        new_H, new_P = Hbeta(knn_dists_row, beta)
+        new_Hdiff = new_H - jnp.log(perplexity)
+        return new_H, new_P, new_Hdiff, new_beta, new_betamin, new_betamax, tries + 1
```
Comment (adamgayoso) [inline src/scib_metrics/utils/_lisi.py:40]: Maybe we should use the flax dataclass (e.g., their trainstate) to represent this

https://flax.readthedocs.io/en/latest/api_reference/flax.struct.html

see the replace method for how to update it inplace

File: src/scib_metrics/_lisi.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    n_cells = knn_graph.shape[0]
+
+    nn_idx = np.empty(shape=(n_cells, n_neighbors), dtype=np.int32)
+    nn_dists = np.empty(shape=(n_cells, n_neighbors), dtype=np.float32)
+    nn_dists[:] = np.NaN
+    for i in range(n_cells):
+        knn_graph_row = knn_graph.getrow(i)
+        sorted_dist_idxs = np.argsort(knn_graph_row.data)
+        nn_idx[i, :] = knn_graph_row.indices[sorted_dist_idxs]
+        nn_dists[i, :] = knn_graph_row.data[sorted_dist_idxs]
```
Comment (adamgayoso) [inline src/scib_metrics/_lisi.py:27]: this can probably be sped up?

Comment (adamgayoso) [reply src/scib_metrics/_lisi.py:27]: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html#sklearn.neighbors.NearestNeighbors

can we use this with precomputed and then call kneighbors?

File: pyproject.toml [hunk 3]
```diff
@@ -21,6 +21,8 @@ urls.Source = "https://github.com/yoseflab/scib-metrics"
 urls.Home-page = "https://github.com/yoseflab/scib-metrics"
 dependencies = [
     "anndata",
+    "flax",
+    "harmonypy",
```
Comment (adamgayoso) [inline pyproject.toml:25]: can you make this a test dependency?

**Review by adamgayoso — CHANGES_REQUESTED** [review]
great, can you also add the lisi stuff to the docs?

File: pyproject.toml (hunk 3 above)
Comment (adamgayoso) [reply pyproject.toml:25]: (harmonypy)

**Comment by adamgayoso** [comment]
also doesn't lisi in scib rescale it somehow?

**Comment by justjhong** [comment]
there is an option to. they have some functions that layer on top which decide for you (labels, batch) but they do (lisi - 1) / (num batches - 1)

**Comment by adamgayoso** [comment]
> there is an option to. they have some functions that layer on top which decide for you (labels, batch) but they do (lisi - 1) / (num batches - 1)

we should add whatever they do

File: src/scib_metrics/utils/_lisi.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+from flax import struct
+
+NdArray = Union[np.ndarray, jnp.ndarray]
+
+
+@struct.dataclass
```
Comment (adamgayoso) [inline src/scib_metrics/utils/_lisi.py:11]: https://github.com/deepmind/chex#dataclass-dataclasspy

Let's actually use this as it's better to depend on chex for this project, as we can use it in the test suite as well

**Review by adamgayoso — CHANGES_REQUESTED** [review]

**Review by adamgayoso — CHANGES_REQUESTED** [review]
Once you add to the docs it LGTM

File: src/scib_metrics/_lisi.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+    nn_obj = NearestNeighbors(n_neighbors=n_neighbors, metric="precomputed").fit(knn_graph)
+    return nn_obj.kneighbors(knn_graph)
+
+
+def lisi_knn(knn_graph: csr_matrix, labels: np.ndarray, perplexity: float = None) -> np.ndarray:
+    """Compute the local inverse simpson index (LISI) for each cell.
+
+    Code inspired by:
+    https://github.com/theislab/scib/blob/e578d84063adf4853ed087500bd3d67078e53337/scib/metrics/lisi.py#L586
```
Comment (adamgayoso) [inline src/scib_metrics/_lisi.py:28]: Instead of this can you add a bib entry for the harmony paper like this?

https://github.com/YosefLab/scib-metrics/blob/a258e0b8bea44549d0225c0e1181b13c556049e2/src/scib_metrics/_silhouette.py#L8

and then the scib paper in the rescaled ones

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: src/scib_metrics/_lisi.py (hunk 5 above)
Comment (justjhong) [reply src/scib_metrics/_lisi.py]: gonna need to commit the other one first for this

**Review by adamgayoso — APPROVED** [review]
